builtin/mainmenu/settings/dlg_file_browser.lua

```diff
@@ -0,0 +1,160 @@
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU Lesser General Public License for more details.
+--
+-- You should have received a copy of the GNU Lesser General Public License along
+-- with this program; if not, write to the Free Software Foundation, Inc.,
+-- 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+local PATH = os.getenv("HOME") -- Linux & similar
+		or os.getenv("HOMEDRIVE") .. os.getenv("HOMEPATH") -- Windows
+		or core.get_user_path() -- If nothing else works
+
+tabdata = {}
+
+local function texture(filename)
+	return core.formspec_escape(defaulttexturedir .. filename .. ".png")
+end
+
+local function get_dirs()
+	local dirs
+
+	if PATH == "\\" then -- Windows root dir
+		dirs = core.get_dir_list(os.getenv("HOMEDRIVE"))
+	else
+		dirs = core.get_dir_list(PATH)
+	end
+
+	local new = {}
+	for _, f in ipairs(dirs) do
+		if f:sub(1,1) ~= "." then
+			new[#new + 1] = core.formspec_escape(f)
+		end
+	end
+
+	table.sort(new)
+	return new
+end
+
+local function make_fs(dialogdata)
+	local dirs = get_dirs()
+
+	local _dirs = ""
+
+	if not next(dirs) then
+		_dirs = "2,,"
+	end
+
+	for _, f in ipairs(dirs) do
+		local is_file = not core.is_dir(PATH .. DIR_DELIM .. f)
+		_dirs = _dirs .. (is_file and "1," or "0,") .. f .. ","
+	end
+
+	local field_label
+	if dialogdata.type == 'file' then
+		field_label = fgettext("Selected file:")
+	elseif dialogdata.type == 'folder' then
+		field_label = fgettext("Selected folder:")
+	end
+
+	return table.concat({
+		"formspec_version[4]",
+		"size[14,9.2]",
+		"image_button[0.2,0.2;0.65,0.65;", texture('up_icon'), ";updir;]",
+		"field[1,0.2;12,0.65;path;;", core.formspec_escape(PATH), "]",
+		"tablecolumns[image,",
+			"0=", texture('folder'), ",",
+			"1=", texture('file'), ",",
+			"2=", texture('blank'),
+			";text]",
+		"table[0.2,1.1;13.6,6;dirs;", _dirs:sub(1,-2),
+			";", (tabdata.selected or 1), "]",
+		"field[0.2,8;9,0.8;select;", field_label, ";", (tabdata.filename or ""), "]",
+		"button[9.5,8;2,0.8;ok;", fgettext("Open"), "]",
+		"button[11.7,8;2,0.8;cancel;", fgettext("Cancel"), "]"
+	})
+end
+
+local function fields_handler(this, fields)
+	if fields.cancel then
+		this:delete()
+		return true
+
+	elseif fields.ok and tabdata.filename and tabdata.filename ~= "" then
+		local is_file = not core.is_dir(PATH .. DIR_DELIM .. tabdata.filename)
+
+		if (this.data.type == "file" and is_file)
+		or (this.data.type == "folder" and not is_file) then
+			core.settings:set(this.data.setting, PATH .. DIR_DELIM .. tabdata.filename)
+
+			this:delete()
+			return true
+		end
+	end
+
+	local dirs = get_dirs()
+
+	if fields.dirs then
+		local event, idx = fields.dirs:sub(1,3), tonumber(fields.dirs:match("%d+"))
+		local filename = dirs[idx]
+		local is_file = not core.is_dir(PATH .. DIR_DELIM .. filename)
+
+		if event == "CHG" then
+			tabdata.filename = filename
+			tabdata.selected = idx
+
+			core.update_formspec(this:get_formspec())
+			return true
```

Comment:
The formspec seems to be automatically updated if the button handler returns `true`, so there is no need for all these `core.update_formspec` calls.